### DISCUSSION TITLE
Make it easier for Mocking

### DISCUSSION THREAD

**Discussion by feinstein** [firebase/flutterfire#3229] [Feature request]
Currently the Firebase architecture uses Singletons which we need to retrieve using `instance` with private constructors, which makes mocking cumbersome.

Copying the tests here is also not great as the Mocks themselves are pretty complicated.

I think the architecture should be rethinked so testing, mocking the Firebase services on our Apps could be easier to implement with Mockito.

As an example, lets look at your own tests [here](https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_auth/firebase_auth/test/firebase_auth_test.dart), you use:

- `setupFirebaseAuthMocks()`
- `MockPlatformInterfaceMixin`

Which are all external references and this makes our apps really hard to Mock. It sure helps your library to be tested, but not so much we as the users to test our own Apps.

**Comment by Ehesp** [comment]
Keen to help here, maybe we could expose a few things internally for testing only. Need to investigate, so any ideas much appreciated.

**Comment by feinstein** [comment]
I would start by creating a demo App that uses most of the services, Auth, Firestore, Crashlytics, trying to use them as the best practices recommend, Dependency Injection, you know, then try to create tests for that App and see how's your experience trying to mock and inject the Firebase services. This way you won't be speculating, but actually seeing the user experience first hand.

For example, you could inject a Repository that uses Firestore internally, so tests would inject a Fake repository (that's easy), but when you want to test the repository itself you would need to inject a fake Firestore instance.

With that you could come up with a more uniform way to mock the services, then the packages could provide some fake implementations under the `/test` folder that we could have access to and streamline the process. Maybe even a separate test package that we include with `dev_dependencies`.

What I suspect you will come up in the end is a more "pluggable" interface, instead of all of these private constructors. You could use the `@visibleForTesting` annotations as well, instead of locking everything.

Also IIRC you construct some fake `FirebaseApp` in your tests, full of fake data, I suspect this is so common you could do it once and reuse it and expose it for us.

I am just thinking out loud here to be honest, but I think this is a step in the right direction and people experienced with tests in Dart should give some hints as well.

It would also be interesting to have a small guideline for tests into the docs.

**Comment by Jonas-Sander** [comment]
I would love to be able to test especially Firestore easily!
